sql-parser/src/main/java/io/crate/sql/Literals.java

```diff
@@ -34,4 +43,47 @@ public static String quoteStringLiteral(String literal) {
     static String escapeStringLiteral(String literal) {
         return literal.replace("'", "''");
     }
+
+    public static String escapeAndQuoteStringLiteral(String literal) {
+        return "E" + Literals.quoteStringLiteral(literal);
+    }
+
+    public static String replaceEscapedChars(String literal) {
+        return replace(
+            replace(
+                replace(
+                    replace(literal,
+                        uni16Pattern,
+                        (s) -> numericStringToString(s, 16)
+                    ),
+                    uni32Pattern,
+                    (s) -> numericStringToString(s, 16)
+                ),
+                hexPattern,
+                (s) -> numericStringToString(s, 16)
+            ),
+            octalPattern,
+            (s) -> numericStringToString(s, 8)
+        );
+    }
+
+    private static String replace(String input, Pattern pattern, UnaryOperator<String> transform) {
+        Matcher matcher = pattern.matcher(input);
+        boolean more = matcher.find();
+        if (!more) {
+            return input;
+        }
+        StringBuffer stringBuffer = new StringBuffer(input.length());
```

Comment:
For single threaded code it would be better to use `StringBuilder`, as it doesn't have synchronization overhead.

The `replaceEscapedChars` logic looks generally a bit expensive as it does multiple pattern matches. I wonder if it would be justified to do a manual loop over the characters and look for the sequences, so we have only one pass?

Reply:
> The `replaceEscapedChars` logic looks generally a bit expensive as it does multiple pattern matches. I wonder if it would be justified to do a manual loop over the characters and look for the sequences, so we have only one pass?

It's something I wanted to discuss further as I also saw it, initially I decided to go for a solution with less complexity. Do you anticipate this to be used heavily?

Reply:
> For single threaded code it would be better to use `StringBuilder`, as it doesn't have synchronization overhead.

in java 1.8 only the StringBuffer variants are supported in Mather class. Should we break bwd compatibility?

Reply:
I think this should still go into 3.3, so no, we shouldn't break compatibility.

I don't think this will be used heavily, but I'm not sure if a more imperative/direct approach is really more complex.

I'd suggest to give it a go and then we can see how it compares.